lte/gateway/python/magma/subscriberdb/client.py

```diff
@@ -77,32 +78,47 @@ def __init__(
         self._grpc_client_manager = grpc_client_manager
 
     async def _run(self) -> None:
-        subscribers = await self._get_subscribers()
-        if subscribers is not None:
-            self._process_subscribers(subscribers)
+        subscribers, no_updates = await self._get_subscribers()
+        # No need to update local db if it's already up-to-date.
+        if subscribers is None or no_updates:
+            return
+        self._process_subscribers(subscribers)
 
     async def _get_subscribers(self) -> [SubscriberData]:
         subscriberdb_cloud_client = self._grpc_client_manager.get_client()
         subscribers = []
         req_page_token = ""  # noqa: S105
         res_page_token = "start_token"  # noqa: S105
+        cur_digest = SubscribersDigest(
+            md5_base64_digest=self._store.get_current_digest(),
+        )
+        no_updates = False
         sync_start = datetime.datetime.now()
-        while res_page_token != "":  # noqa: S105
+
+        # If response.no_updates = True, it means local data is up-to-date &
+        # therefore no need for updates from the cloud.
+        while res_page_token != "" and not no_updates:  # noqa: S105
```

Comment:
Update the cloud such that it sends an empty page token when there are no updates

The canonical way of reading all subs should be, simply, read till returned page token is empty. no_updates tells you what to do *after* you finish reading all subs, not whether you're done reading

```suggestion
        while res_page_token != "":  # noqa: S105
```

Reply:
P.S. I prefer the "while not found_empty_token" construct we used in the cloud servicer tests https://github.com/magma/magma/pull/7302/files#diff-3ac163790ecd8ba19e710823f9528dec68bd82b18f0de71cad82c1066fc3d0c2R131-R133